### ISSUE TITLE
ResourceManager.GetString always returns english when using netcoreapp

### ISSUE THREAD

**Issue by ryanbrandenburg** [dotnet/corefx#12625]
This is a continuation of aspnet/Localization#293.
## Steps to reproduce
1. Clone and run [this](https://github.com/ryanbrandenburg/ResourceManagerNetCoreApp) repo app.
2. See that the value from the english resx has been returned even though the current culture is swedish.
3. Move `net451` to the top of the frameworks in project.json to run the project under `net451` instead of netcoreapp.
4. Note that when run under `net451` an exception is thrown because the swedish version of the given resource doesn't exist.

Note that this only seems to apply to english (at least on my computer, likely this is dependent on the installed language), if you only had a french resource in your project you would get the expected behavior of a throw exception.
## Expected behavior

We expect the  `netcoreapp` version of `ResourceManager` to behave the same as the `net451` one, throwing an exception when a resource isn't found and no neutral language is explicitly set.
## Actual behavior

The english resource is returned, even though the current culture is swedish.

**Comment by ryanbrandenburg** [comment]
Interestingly setting the NeutralResourcesLanguage in AssemblyInfo DOES set resourceManager._neutralResourcesCulture to that language, but does NOT affect this behavior as one might expect after reading [this](https://github.com/dotnet/corefx/blob/master/src/System.Resources.ResourceManager/src/System/Resources/ResourceManager.cs#L97) area of the ResourceManager.

**Comment by ramarag** [comment]
@tarekgh  the divergence in behavior was caused by your change in 
https://github.com/dotnet/corefx/commit/ba6f2955580a029037d920c4be836e9850427e46

comments on how to proceed further?

**Comment by tarekgh** [comment]
@ramarag you are looking at wrong place as the change you referred to is AOT specific and not used in the scenario we are talking about.

**Comment by gkhanna79** [comment]
@ryanbrandenburg Does this work as expected in .NET Core 1.0.0? I am not aware of any ResourceManager specific work in .NET Core 1.1.0, so I am trying to understand if this is a regression or not.

**Comment by ajtowf** [comment]
@gkhanna79 didn't work with 1.0.1, I don't know for 1.0.0 but my guess is that it doesn't work for 1.0.0 either. But then again, that's _very_ easy to test by cloning the example repo...

**Comment by ramarag** [comment]
Summary:
 
The divergence in behavior in desktop and .net core is due to a Feature that supports `OS Fallback for looking up Resources` that was enabled only on coreclr from the time of Silverlight.
If we are to make the resource iteration logic compatible with desktop, then this would be a breaking change from .netcore 1.0
 
Details:
 
On desktop, if we are not able to lookup resources by walking up the parent chain of the cultures, we finally return the invariant culture. In the current scenario, the Invariant culture is not set and hence we throw the MissingManifestResourceException exception 
 
On Coreclr, since the OS Fallback is enabled, we walk the fall back chain before looking up the invariant culture. Here the  OS provides “en-us and en” as the fallback culture , as there are resources available for “en” , coreclr returns the resources for “en”

**Comment by ramarag** [comment]
The implementation divergence is at https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Resources/ResourceFallbackManager.cs#L95

**Comment by ajtowf** [comment]
So what is the recommendation here? Could I just set my default language that I use in the views to be the invariant culture for the assembly as well and life is good?

**Comment by tarekgh** [comment]
@ajtowf what exactly the scenario you are talking about here? the repro sample is just returning the English resource when requesting resources with a language which not exist in the app. if the app has resources for this language it should be returned and you shouldn't get the English resources. 

the only difference I am seeing compared to the desktop, in case not finding the resources coreclr will return the default resource matching the underlying OS language if exist. and on desktop we'll throw. 

could you please clarify more about the scenario you want handle?

**Comment by ajtowf** [comment]
@tarekgh it's the scenario that originated this bug, https://github.com/aspnet/Localization/issues/293 it's very basic and it's the scenario that's described in the official documentation!

We have a view, the app is swedish (välkommen means welcome):
@Localizer["Välkommen"]

The user can switch to english, so there are only english resx files, since the default swedish language is in the views. Now the user will _always_ get the english translation even for the default language which is swedish.

It's very frustrating to need to explain this problem over and over again, excuse my frustrated tone, but please read the comments in the bug.

**Comment by tarekgh** [comment]
@ajtowf @ryanbrandenburg 

we are talking about 2 layers here, one layer is the ASP.NET localization model, the second layer is the net core localization layer.
my question was mainly concerning the second layer (net core localization). what I am trying to do is to isolate the problem. my question is, if you have simple net core app (not ASP.NET) and you have localized Swedish resources in this app and then at run time you set the current UI culture to Swedish and then called ResourceManager.GetString. do you still get the English resources? from my understanding from Rama this is not happening and we should be getting the Swedish resources. if this is the case, then the problem would be in the ASP.NET layer.

@ajtowf sorry for your frustration but we are really trying to help.

**Comment by ajtowf** [comment]
@tarekgh right, and from what I can see [the sample repo](https://github.com/ryanbrandenburg/ResourceManagerNetCoreApp) created by @ryanbrandenburg in this thread does exactly that, i.e. using the net core localization layer _without_ aspnet, and outputting:

    Hello from resx

I guess I find it confusing that we're getting questions that is answered by simply running the sample repo. Or am I missing something obvious here?

**Comment by gkhanna79** [comment]
In the basic repro, as @ramarag explained, the app comes with "en" resources and since we cannot find the Swedish resources, we ask the OS for a fallback culture list in netcore. This returns "en" and since we find the resources for the same, we return them. If "en" resources were not present, then you would get an exception as well.

Building further, what is application intending to do on net451 when the resources are not found and the exception is thrown?

With the above said, we intend to look into this for our next major release to determine the implications of unifying the behavior across runtimes and make appropriate changes.

**Comment by tarekgh** [comment]
> I guess I find it confusing that we're getting questions that is answered by simply running the sample repo. Or am I missing something obvious here?

as @gkhanna79 mentioned when you compile the repro app we get English resources only. so there is no Swedish resources in the repro app. did you run the sample app yourself and see it has Swedish resources? or you just guessing this was the case?

**Comment by ajtowf** [comment]
@gkhanna79 cool, thanks!

**Comment by ajtowf** [comment]
@tarekgh what do you think mate? I'm aware there's no swedish words in the sample app obviously, but we should see "Hello" instead of "Hello from resx"...

![image](https://cloud.githubusercontent.com/assets/3427270/19898161/2757c474-a05b-11e6-9424-1376fb245bb1.png)

**Comment by ajtowf** [comment]
@gkhanna79 what would you say is the recommended work around for the moment? To use const keys that we pass to the resource manager and always have resx files for all supported languages? Meaning no "default" language in the AspNet Razor View files. Right now, I'm putting the English translation into Norwegian resx-files to get around this.

**Comment by tarekgh** [comment]
@ajtowf 

this is happening because that what has been specified in the resx file

```XML
<data name="Hello" xml:space="preserve">
    <value>Hello from resx</value>
  </data>
```

I hope this can reduce the confusion.

**Comment by gkhanna79** [comment]
@ajtowf ResourceManager.Getstring returns the value for the key you pass in the call to GetString (see https://msdn.microsoft.com/en-us/library/d17ax2xk(v=vs.110).aspx). Thus, you are returned the value from https://github.com/ryanbrandenburg/ResourceManagerNetCoreApp/blob/master/src/ResourceManagerNetCoreApp/Resources/Program.en.resx#L121.

**Comment by ajtowf** [comment]
@tarekgh right, that's in the Program**.en.**resx file, and I don't expect to get that value if the culture is set to **sv**. According to the following documentation I should get "Hello" since the culture is set to something different than "**en**":

https://docs.asp.net/en/latest/fundamentals/localization.html#view-localization and
https://docs.asp.net/en/latest/fundamentals/localization.html#provide-localized-resources-for-the-languages-and-cultures-you-support

Can we at least agree on that?

**Comment by tarekgh** [comment]
@ajtowf now we are switching to the first layer I talked about previously (which is APS Localization model). The issue as @gkhanna79 and @ramarag described before is in net core, the resource manager fallback to default UI language on your system (which is English) so you end up with English resources. in other word, the resource manager returned some value. I believe ASP.NET just return that value to the app. 
on the desktop, the behavior is different as the desktop don't fallback to default UI language on your system and at that time net core resource manager not returning any value and then ASP.NET (I think) will just return the key you have passed "Hello". 

I think the work around you can do (if possible) is either to remove the English resources from your app  or add Swedish resources to the app. I'll try to look more if we can find other possible work around.

**Comment by karelz** [comment]
Next step: Should be simple to remove the code and throw the same way as on Desktop.
See hint in: https://github.com/dotnet/corefx/issues/12625#issuecomment-257447037

We still may need to introduce compat switch to support behavior of .NET Core 1.0 and 1.1.

**Comment by Eilon** [comment]
@karelz we have ASP.NET Core scenarios that are blocked by this. Can we expect this bug to be fixed for the first preview of 2.0?

cc @DamianEdwards

**Comment by karelz** [comment]
@ramarag @tarekgh can you please prioritize this fix to unblock ASP.NET team?
cc: @danmosemsft @gkhanna79

**Comment by jkotas** [comment]
cc @morganbr

**Comment by gkhanna79** [comment]
We will get this issue fixed for the first ASP.NET Core 2.0 Preview once we are done with our impending feature work.

**Comment by DamianEdwards** [comment]
@gkhanna79 this is less about the first preview (which is a while away) and more about unblocking work right now. Can you clarify the timelines you're eluding to please?

**Comment by gkhanna79** [comment]
@ramarag would be able to look into this sometime next week.

**Comment by ryanbrandenburg** [comment]
Can this be closed now that dotnet/coreclr#9472 is merged?
